c2corg_api/views/sso.py

```diff
@@ -0,0 +1,282 @@
+from base64 import b64encode
+from datetime import datetime, timedelta
+from os import urandom
+from pytz import utc
+from urllib.parse import urlencode
+import logging
+
+import colander
+from cornice.resource import resource, view
+from cornice.validators import colander_body_validator
+from pydiscourse.exceptions import DiscourseClientError
+from pyramid.httpexceptions import (
+    HTTPForbidden,
+    HTTPFound,
+    HTTPInternalServerError,
+)
+
+from c2corg_common.attributes import default_langs
+
+from c2corg_api.models import DBSession
+from c2corg_api.models.document import DocumentLocale
+from c2corg_api.models.sso import (
+    SsoKey,
+    SsoExternalId,
+)
+from c2corg_api.models.user import User
+from c2corg_api.models.user_profile import UserProfile
+from c2corg_api.security.discourse_client import get_discourse_client
+from c2corg_api.security.roles import log_validated_user_i_know_what_i_do
+from c2corg_api.views import (
+    cors_policy,
+    json_view,
+)
+from c2corg_api.views.user import (
+    validate_unique_attribute,
+    validate_forum_username,
+)
+
+CONST_EXPIRE_AFTER_MINUTES = 10
+
+log = logging.getLogger(__name__)
+
+
+class SsoSyncSchema(colander.MappingSchema):
+    sso_key = colander.SchemaNode(colander.String())
+    external_id = colander.SchemaNode(colander.String(),
+                                      missing=None),
+    email = colander.SchemaNode(colander.String(),
+                                missing=None),
+    username = colander.SchemaNode(colander.String(),
+                                   missing=None)
+    name = colander.SchemaNode(colander.String(),
+                               missing=None)
+    forum_username = colander.SchemaNode(colander.String(),
+                                         missing=None)
+    lang = colander.SchemaNode(colander.String(),
+                               validator=colander.OneOf(default_langs),
+                               missing=None)
+    groups = colander.SchemaNode(colander.String(),
+                                 missing=None)
+
+sso_sync_schema = SsoSyncSchema()
+
+
+def sso_sync_validator(request, **kwargs):
+    if 'sso_key' not in request.json:
```

Comment:
Shouldn't you rather manipulate ``request.validated`` instead of ``request.json``? See for instance https://github.com/c2corg/v6_api/blob/master/c2corg_api/views/document_merge.py#L29